### DISCUSSION TITLE
Make it easier to share rules

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#3829] [Ideas]
Lately I see some good proposals as rules but I think that they don't fit in our default rule sets. We always say: "write a custom rule". But nearly no one publish it.

We need a way to make it even easier for the users to share their rules. The easiest way I can think of is to create a [gist](https://gist.github.com/) with a rule and add the url in the `yaml`. Could that be possible? Which other ideas do you have to make this task easier?

**Comment by JavierSegoviaCordoba** [comment]
That approach is indexable?

**Reply by BraisGabin** [reply]
To be honest, I have no idea if it is even doable.

But, what do you mean with indexable? A repository where you could find all of them? I imagine that we could create a web, wiki or something similar.

**Comment by cortinico** [comment]
> Could that be possible? Which other ideas do you have to make this task easier?

I think we should add a page to the website where contributors can easily add their custom rules. 
Ideally for every rule/ruleset we want to list:

* Rule name
* Repository Link
* Maven repo where is it hosted
* Gradle line to add this rule to their project

**Reply by BraisGabin** [reply]
> Maven repo where is it hosted

I think that this creates a lot of friction. Would it be possible to avoid this?

**Reply by cortinico** [reply]
I actually thought the opposite 🤔 If you have your custom rule in a github repo but I can't get the binary out of it, that adds a lot of frictions on the consumer end. Like you're forcing the consumer to copy-n-paste the code or build the rules locally.

As a last resort jitpack also works for this kind of solution.

**Reply by BraisGabin** [reply]
You are talking about the friction for the person that want to use it I'm talking about the friction of the person that wants to share it. Both frictions should be lowered for sure, but the problem that I see right now is that there is not rules out there to peak. Why is that? Upload something to maven central is a pain. It's really secure and those things but if I want to share a rule I don't want to do that.

My idea was more "out of the box". Kotlin can work as a script language so: could I upload a rule to a gist, add the url in the config yaml and run it? I think that it should have really little fricton for both creator and consumer.

Something like:

```yaml
remote:
  - https://gist.github.com/BraisGabin/8958c44f3cf9e939294e25e82547862c # this is not a rule, it is a random gist I have
    active: true
```

Right now detekt is far from allowing this. But it is doable? Can you think of another way to make rules easy to share?

I know that this option that I'm giving is REALLY insecure. Anyone could create a cool rule, get a lot of users and then change the code to do whatever he want. We would need something for the versioning. But as I told before: how could we make the rules easier to share?

**Reply by cortinico** [reply]
> Right now detekt is far from allowing this. But it is doable? Can you think of another way to make rules easy to share?

Imho the gist approach is far from optimal (how do you compile it? how do you list dependencies if needed, etc). I would opt for maven-hosted solution. I agree that publishing to Maven Central is a pain. But we can still fallback to using https://jitpack.io/ that requires no customization.

What I think we can do is create a Github Template repository, say `detekt/detekt-custom-rule`. With a minimal Github Actions setup and an empty rule. Contributors can just use that template and start adding their custom logic.

That should work out of the box with Jitpack. Contributors can go the extra mile and setup the publishing to Central or wherever they wish.

**Reply by JavierSegoviaCordoba** [reply]
The problem about publishing is the Maven central setup (domain, jira, etc).

What about detekt provides a published plugin which allow upload easily rules to GitHub packages or something so?

Or even maybe just creating s PR to detekt repo and just publish them under detekt domain.

**Reply by cortinico** [reply]
> What about detekt provides a published plugin which allow upload easily rules to GitHub packages or something so?

Yeah that's why I mentioned Jitpack. Github packages requires to authenticate with Github to consume artifacts, making it really unpractical.

**Reply by BraisGabin** [reply]
I like that! A template and jitpack could help a lot.
